File: 코드/ch6/clip1/boj1987/main.py
import sys

def solve(row, col):
    result = 0
    for i in range(4):
        nr, nc = row + dr[i], col + dc[i]
        if is_out_of_bound(nr, nc, r, c):
            continue
        
        next_char = board[nr][nc]
        route = 1 << next_char
        if visited[nr][nc] == (visited[row][col] | route):
            continue
        if check[next_char]:
            continue
        
        visited[nr][nc] = visited[row][col] | route
        check[next_char] = True
        result = max(result, solve(nr, nc))
        check[next_char] = False
    return result + 1

def is_out_of_bound(row, col, bound_r, bound_c):
    return row < 0 or row >= bound_r or col < 0 or col >= bound_c

r, c = map(int, sys.stdin.readline().split())
board = []
for _ in range(r):
    line = sys.stdin.readline().strip()
    board.append([ord(ch) - ord('A') for ch in line])

dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

visited = [[0] * c for _ in range(r)]
check = [False] * 26
check[board[0][0]] = True
visited[0][0] = 1 << (board[0][0])
print(solve(0, 0))

File: 코드/ch6/clip1/boj1987/Main.java
import java.util.Scanner;

class Main {
    public static int[][] board;
    public static boolean[] check = new boolean[26];
    public static int r, c;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        r = sc.nextInt();
        c = sc.nextInt();
        board = new int[r][c];
        visited = new int[r][c];
        for (int i = 0; i < r; i++) {
            String line = sc.next();
            for (int j = 0; j < c; j++) {
                board[i][j] = line.charAt(j) - 'A';
            }
        }
        visited[0][0] = 1 << (board[0][0]);
        check[board[0][0]] = true;
        System.out.println(solve(0, 0));
    }

    // dr[0] dc[0] : 상
    // dr[1] dc[1] : 우
    // dr[2] dc[2] : 하
    // dr[3] dc[3] : 좌
    static int[] dr = {-1, 0, 1, 0};
    static int[] dc = {0, 1, 0, -1};

    // visited[row][col] -> (row, col) 도달했을때 사용한 알파벳
    public static int[][] visited;

    public static int solve(int row, int col) {
        int result = 0;
        for (int i = 0; i < 4; i++) {
            int nr = row + dr[i], nc = col + dc[i];
            if (isOutOfBound(nr, nc, r, c)) continue;
            int next = board[nr][nc];
            if (check[next]) continue;
            int route = 1 << next;
            if (visited[nr][nc] == (visited[row][col] | route)) continue;

            visited[nr][nc] = visited[row][col] | route;
            check[next] = true;
            int nextResult = solve(nr, nc);
            result = Math.max(result, nextResult);
            check[next] = false;
        }
        return result + 1;
    }

    static boolean isOutOfBound(int row, int col, int boundR, int boundC) {
        return row < 0 || col < 0 || row >= boundR || col >= boundC;
    }
}